xfer/collector.go

```diff
@@ -1,233 +1,59 @@
-	initialBackoff = 2 * time.Second
-)
-
-var (
-	// MaxBackoff is the maximum time between connect retries.
-	// It's exported so it's externally configurable.
-	MaxBackoff = 1 * time.Minute
-
-	// This is extracted out for mocking.
-	tick = time.Tick
-)
-
-// Collector describes anything that can have addresses added and removed, and
-// which produces reports that represent aggregate reports from all collected
-// addresses.
-type Collector interface {
-	Add(string)
-	Remove(string)
-	Reports() <-chan report.Report
-	Stop()
-}
-
-// realCollector connects to probes over TCP and merges reports published by those
-// probes into a single one.
-type realCollector struct {
-	in     chan report.Report
-	out    chan report.Report
-	peekc  chan chan report.Report
-	add    chan string
-	remove chan string
-	quit   chan struct{}
-	id     string
+// Collector receives published reports from multiple producers. It yields a
+// single merged report, representing all collected reports.
+type Collector struct {
+	mtx     sync.Mutex
+	reports []timestampReport
+	window  time.Duration
 }
 
-// NewCollector produces and returns a report collector.
-func NewCollector(batchTime time.Duration, id string) Collector {
-	c := &realCollector{
-		in:     make(chan report.Report),
-		out:    make(chan report.Report),
-		peekc:  make(chan chan report.Report),
-		add:    make(chan string),
-		remove: make(chan string),
-		quit:   make(chan struct{}),
-		id:     id,
+// NewCollector returns a collector ready for use.
+func NewCollector(window time.Duration) *Collector {
+	return &Collector{
+		window: window,
 	}
-	go c.loop(batchTime)
-	return c
 }
 
-func (c *realCollector) loop(batchTime time.Duration) {
-	var (
-		tick    = tick(batchTime)
-		current = report.MakeReport()
-		addrs   = map[string]chan struct{}{}
-		wg      = &sync.WaitGroup{} // per-address goroutines
-	)
-
-	add := func(ip string) {
-		if _, ok := addrs[ip]; ok {
-			return
-		}
-		addrs[ip] = make(chan struct{})
-		wg.Add(1)
-		go func(quit chan struct{}) {
-			defer wg.Done()
-			c.reportCollector(ip, quit)
-		}(addrs[ip])
-	}
-
-	remove := func(ip string) {
-		q, ok := addrs[ip]
-		if !ok {
-			return // hmm
-		}
-		close(q)
-		delete(addrs, ip)
-	}
-
-	for {
-		select {
-		case <-tick:
-			c.out <- current
-			current = report.MakeReport()
-
-		case pc := <-c.peekc:
-			copy := report.MakeReport()
-			copy.Merge(current)
-			pc <- copy
-
-		case r := <-c.in:
-			if err := r.Validate(); err != nil {
-				log.Printf("Received invalid report: %v", err)
-				continue
-			}
-			current.Merge(r)
+var now = time.Now
 
-		case ip := <-c.add:
-			add(ip)
+// Add adds a report to the collector's internal state.
+func (c *Collector) Add(rpt report.Report) {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	c.reports = append(c.reports, timestampReport{now(), rpt})
+}
 
-		case ip := <-c.remove:
-			remove(ip)
+// Report returns a merged report over all added reports.
+func (c *Collector) Report() report.Report {
```

Comment:
Is something periodically calling Report, or is it only called when a UI is attached?

If so, won't c.reports get arbitrarily long?